### PR TITLE
Fix for init container cpu leak in kubelet cpu manager's static policy

### PR DESCRIPTION
**Pull request by hemanthmalla**
When using [static cpu policy](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/#static-policy) for containers, kubelet's cpu manager tries to reuse cpus previously  allocated to init containers for app containers. However cpu manager tries to allocate all the requested cpus from the same core if possible. Currently allocation from the same core is preferred over reusing init container cpus. When this happens, cpus not reused from init containers are never released back to the default cpuset. 

This results in leaking cpus and  kubelet rejecting workloads with `UnexpectedAdmissionError`. Since the scheduler isn't aware of this, this can go on in a loop 🔁 , creating hundreds of pods that end up in `UnexpectedAdmissionError` state.

Here's an example scenario on a machine with 8 cpus and 4 cores:
```
Pod 1 
init  ctr 1 : 2 cpus (0, 4) cpus to reuse [0, 4]
init  ctr 2 : 4 cpus (0-1, 4-5) cpus to reuse [0, 1,  4, 5]
app ctr 1 : 3 cpus (0-1, 4) cpus to reuse [5]
app ctr 2 : 2 cpus (2-6) 

For app container 2, 2-6 cpus are allocated since they're on the same core. Now cpu 5 is never released back to default pool.
Now, we don't have enough cpus for pod 2.

Pod 2
init  ctr 1 : 3 cpus ()
app ctr 1 : 3 cpus ()
```

Credits to @nyodas for finding the pattern to reproduce the bug.

### CODE REVIEW

File: pkg/kubelet/cm/cpumanager/policy_static.go [hunk 1]
```diff
@@ -238,7 +232,16 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
 		}
 		s.SetCPUSet(string(pod.UID), container.Name, cpuset)
 		p.updateCPUsToReuse(pod, container, cpuset)
+		// If the recently allocated CPU is for the last app container in the pod, release CPUs that are not re-used
+		// back to the default CPU set.
+		if container.Name == pod.Spec.Containers[len(pod.Spec.Containers)-1].Name{
+			cpusToReturn, ok := p.cpusToReuse[string(pod.UID)]
+			if  ok && ! cpusToReturn.IsEmpty(){
```
Comment (EricMountain) [inline pkg/kubelet/cm/cpumanager/policy_static.go:239]: Nit: few spacing issues. I suspect gofmt hasn't run?

```suggestion
			if ok && ! cpusToReturn.IsEmpty() {
```

File: pkg/kubelet/cm/cpumanager/cpu_manager_test.go [hunk 2]
```diff
@@ -533,16 +534,157 @@ func TestCPUManagerAddWithInitContainers(t *testing.T) {
...
+		initContainerIDs [][]string
+		containerIDs     [][]string
+		stAssignments    state.ContainerCPUAssignments
+		stDefaultCPUSet  cpuset.CPUSet
+		pod              []*v1.Pod
+		expInitCSets     [][]cpuset.CPUSet
+		expCSets         [][]cpuset.CPUSet
+	}{
+		{
+			description:     "Multi Init, Multi App Container Split CPUs",
```
Comment (EricMountain) [inline pkg/kubelet/cm/cpumanager/cpu_manager_test.go:564]: We should change this because another test has the same description l.457.

**Review by nyodas — APPROVED** [review]

File: pkg/kubelet/cm/cpumanager/policy_static.go [hunk 3]
```diff
@@ -238,7 +232,16 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
...
+				s.SetDefaultCPUSet(s.GetDefaultCPUSet().Union(cpusToReturn))
+				delete(p.cpusToReuse, string(pod.UID))
+			}
```
Comment (nyodas) [inline pkg/kubelet/cm/cpumanager/policy_static.go:243]:
```suggestion
```

File: pkg/kubelet/cm/cpumanager/cpu_manager_test.go [hunk 4]
```diff
@@ -533,16 +534,157 @@ func TestCPUManagerAddWithInitContainers(t *testing.T) {
...
+			containerIDs := append(
+				testCase.initContainerIDs[ix],
+				testCase.containerIDs[ix]...)
+
+			expCSets := append(
+				testCase.expInitCSets[ix],
+				testCase.expCSets[ix]...)
+
+			for i := range containers {
+				err := mgr.Allocate(pod, &containers[i])
```
Comment (EricMountain) [inline pkg/kubelet/cm/cpumanager/cpu_manager_test.go:651]: Something I don’t quite follow regarding how this works:

Init containers are allocated CPUs by the call to mgr.Allocate(), correct?

How do the CPUs allocated to init containers get “released” so they can be allocated to main containers? I don’t see an explicit call that does that. Is there something implicit happening here? Just trying to understand better how the test works.

**Pull request closed without merging**